EIPS/eip-00.md

```diff
@@ -0,0 +1,831 @@
+
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
+
+```
+
+Same as in the [`sender` validation frame](#sender-validation-frame), in order to support gas estimation this
+frame should return a value different from `MAGIC_VALUE_PAYMASTER` for conditions that cannot be satisfied
+before signing.
+
+The size of the `context` byte array may not exceed `MAX_CONTEXT_SIZE` for a transaction to be considered valid.
+
+#### Sender execution frame
+
+The `sender` address is invoked with `callData` input.
+
+The gas limit of this frame is set to `callGasLimit - calldataCost`.\
+Calculation of the `calldataCost` value is defined in the
+[Gas fees charged for transaction input](#gas-fees-charged-for-transaction-input) section.\
+The amount of gas used by this frame is referred to as `gasUsedByExecution`.
+
+The validation frames do not revert if the execution frame reverts.
+The `postPaymasterTransaction` may still be called with a `success: false` flag.
+
+#### Paymaster post-transaction frame
+
+After the sender execution frame is over the `paymaster` may need to perform some post-transaction logic,
+for instance to perform some kind of cleanup or bookkeeping.
+If the gas payment validation returned a non-zero `context`, the `paymaster` is invoked again
+with the following inputs:
+
+```solidity
+
+function postPaymasterTransaction(bool success, uint256 actualGasCost, bytes context) external;
+
+```
+
+The `actualGasCost` parameter is the actual amount paid by the paymaster for this transaction,
+and `success` indicates whether this transaction's execution frame completed without revert.
+
+The gas limit of this frame is set to `paymasterGasLimit - paymasterValidationGasUsed`.
+
+Revert in the `postPaymasterTransaction` frame reverts the transaction's execution frame as well.
+The validation frames do not revert if the `postPaymasterTransaction` frame reverts.
+The gas fees charged from the `paymaster` will still include the gas cost of the reverted execution frame.
+
+### Execution flow diagram
+
+The execution flow determined by an Account Abstraction Transaction is visualised by the following flow diagram:
+
+![](../assets/eip-00/flow_diagram.png)
+*Execution flow for the Native Account Abstraction Transactions*
+
+### Execution layer transaction validation
+
+On the execution layer, the transaction validity conditions for a block are extended as follows:
+
+```
+
+func validateAccountAbstractionTransaction(tx *Transaction) {
+    assert !(sender.code.length > 0 && deployerData.length > 0)
+
+    if (sender.code.length == 0 && deployerData.length == 0) {
+        validUntil = (nonce >> 112) & 0xffffffffffff
+        validAfter = (nonce >> 160) & 0xffffffffffff
+        assert Date.now() <= validUntil
+        assert Date.now() >= validAfter
+    }
+
+    if (sender.code.length == 0 && deployerData.length > 0) {
+        assert deployerData.length > 20
```

Comment:
Technically we could allow a deployer with empty calldata, but is there a safe way to use such deployer?  The deployed address won't be associated with the sender, so any deployment transaction could be frontrun to grab the address before the user does.

But I agree, if you see a safe way to implement a zero-calldata deployer then we should support it.

Reply:
Sender address is derived from deployer and deployerData.
So empty deployerData means it deploys a singleton.
It's a one-time-use deployer.

Reply:
The address is determined by the factory contract, not by the protocol.  The natural way for the factory to do it, is to use deployerData to set the CREATE2 address.  But technically it could get the data by calling some contract, or just using a nonce and behave like CREATE.  I can't think of a good use for such factory, as frontrunners would be able to grab the sender's address and invalidate the transaction, but do we have a reason not to allow it.